consensus/hotstuff/vote_collector.go

```diff
@@ -0,0 +1,28 @@
+package hotstuff
+
+import (
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// VoteCollector collects votes for the same block, produces QC when enough votes are collected
+// VoteCollector takes a callback function to report the event that a QC has been produced.
+type VoteCollector interface {
+	// AddVote adds a vote to the collector
+	// returns true if the vote was added
+	// returns false otherwise
+	// return error if the signature is invalid
+	// When enough votes have been added to produce a QC, the QC will be created asynchronously, and
+	// passed to EventLoop through a callback.
+	AddVote(vote *model.Vote) (bool, error)
+
+	// VoteCreator returns a function that is able to create a vote for a given block
+	// The returned createVote function holds a stateful crypto signer object that is able to
+	// verify and store a signature share from other nodes, as well as produce a signature for
+	// our own vote. It is used by Voter to create our own vote.
+	VoteCreator() createVote
```

Comment:
This is for the case where we receive the block _after_ receiving votes. 

Let's go through this case:

1. The compliance engine received a vote, it forwards it to the VoteAggregator
2. VoteAggregator adds this vote to vote queue
3. Vote Queue worker takes this vote out of the queue, and asks the VoteCollectors: do we have a vote collector for the block this vote is voting for.
3. The VoteCollectors says no. Then the worker creates a CachingVoteCollector, add to the vote collectors map, and asks it to add (hold) this pending vote.
4. Now compliance receives the block of the vote. It forwards to EventHandler
5. EventHandler checks the block's QC, good, then gives it to VoteAggregator.
6. VoteAggregator asks VoteCollectors: do we have a vote collector for this block? Yes, and it return the cachingVoteCollector after acquiring the read lock(Here we need a way to tell VoteAggregator what's the status of the VoteCollector, so that VoteAggregator can know this is a cachingVoteCollector, we probably need to add a `Status() VoteCollectorStatus` method to VoteCollector), and it will create a new VerifyingVoteCollector, use it to verify the block and add the proposer vote, then replace the cachingVoteCollector with the VerifyingVoteCollector in the map, and return the validation result back to the EventHandler.
7. EventHandler learned the block is valid, it asks Voter whether to vote for this block. 
8. Voter checked the block, and say it decides to vote, it uses the `BlockSigner` struct (I mentioned above) as a bridge create a vote, which will first access the VoteCollectors map, and read back the VoteCollector interface, now we can check its state and cast it into VerifyingVoteCollector, because EventHandler must have add this block to VoteAggregator before. And the VerifyingVoteCollector could provide this `VoteCreator` method for the BlockSigner object to create the Vote.

This is why VoteCreator method is needed.